### PR TITLE
[Feature Fix]Public components not showing on Profile

### PR DESCRIPTION
**Pull request by GageGaskins**
# Purpose
Solves #3141. Render_nodes.mako is called twice and render_nodes.js was using the id to applyBindings. This caused an error with Knockout for attempting to apply bindings on the same element twice, leading to Public Components not being rendered at all.

# Changes
Added a check to osfHelpers applyBindings to only apply bindings if the current element hasn't done so already. Added logic to find any componentScope class in render_nodes.js so that all node lists have bindings applied properly.

# Side Effects
There shouldn't be any side effects as this was so far the only case where Knockout bindings were being applied more than once, which should be fixed for any case now with the check, and the componentScope class is only mentioned within the render_nodes.mako file.d

### CODE REVIEW

File: website/static/js/pages/render-nodes.js [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
+var $ = require('jquery');
 
 // model for components, due to simplicity did not create a new file
 var ComponentControl = {};
 
 // binds to component scope in render_nodes.mako
-$osf.applyBindings(ComponentControl, '#componentScope');
+$(document).ready(function() {
+    $('.componentScope').each(function(i) {
+       $(this).attr('id', 'componentScope' + i);
```
Comment (lyndsysimon) [inline website/static/js/pages/render-nodes.js]: Why is a unique ID necessary?

File: website/static/js/pages/render-nodes.js [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
 
 var $osf = require('js/osfHelpers');
+var $ = require('jquery');
 
 // model for components, due to simplicity did not create a new file
 var ComponentControl = {};
 
 // binds to component scope in render_nodes.mako
-$osf.applyBindings(ComponentControl, '#componentScope');
+$(document).ready(function() {
```
Comment (lyndsysimon) [inline website/static/js/pages/render-nodes.js]: Why does this now require delaying until the document is loaded, but didn't prior to the change?

File: website/templates/util/render_nodes.mako [hunk 3]
```diff
@@ -1,6 +1,6 @@
 % if len(nodes):
     <ul class="list-group ${'sortable' if sortable and 'write' in user['permissions'] else ''}">
-      <span id="componentScope" class="scripted">
+      <span class="componentScope scripted">
```
Comment (lyndsysimon) [inline website/templates/util/render_nodes.mako]: Our style guide says that classes and IDs follow different naming conventions. This class should be 'component-scope'.

I'd also strongly consider renaming it. `component-scope` isn't very description - perhaps `node-list-scope` or `rendered-nodes-scope`?

File: website/templates/util/render_nodes.mako [hunk 4]
```diff
@@ -46,7 +46,7 @@
             <a href="https://osf.io/getting-started/#privacy" target="_blank">public</a>.
         </p>
     </div>
-% elif profile is not UNDEFINED:  # On profile page and user has no public projects/components
+% elif profile is not UNDEFINED:  ## On profile page and user has no public projects/components
```
Comment (lyndsysimon) [inline website/templates/util/render_nodes.mako]: What's the purpose of this change?

File: website/static/js/osfHelpers.js [hunk 5]
```diff
@@ -456,14 +457,19 @@ var applyBindings = function(viewModel, selector) {
...
     // Also show any child elements that have the scripted class
-    $(cssSelector + ' .scripted').each(function(elm) {
+    $(cssSelector + ' .scripted').each(function (elm) {
         $(this).show();
     });
-    ko.applyBindings(viewModel, $elem[0]);
+
+    //Check if bindings have been applied to the element already
+    var isBound = !!ko.dataFor(elem);
+    if(!isBound) {
```
Comment (lyndsysimon) [inline website/static/js/osfHelpers.js]: I would prefer this not be included - the warning that you are attempting to bind an object twice is useful in that it tells the developer that something is wrong. I'm concerned that allowing that case to pass silently could complicate debugging.

**Comment by lyndsysimon** [comment]
First pass complete.

File: website/templates/util/render_nodes.mako (hunk 4 above)
Comment (GageGaskins) [inline website/templates/util/render_nodes.mako]: I think this was just a typo fix I forgot to mention.

File: website/static/js/osfHelpers.js (hunk 5 above)
Comment (GageGaskins) [inline website/static/js/osfHelpers.js]: I agree, but this is currently the only fix I could come up with without having to rewrite the logic of how profile page is rendered. What happens is that public projects calls render_nodes.mako, which when finished calls render_nodes.js, and it applies its bindings. However, when public components calls render_nodes.mako, it will find the public project nodes, try to reapply the bindings, and fail. Maybe I could attempt another fix entirely contained in render_nodes.js, but as of now because render_nodes.mako is called twice, anything within render_nodes.js will be called twice, specifically applyBindings.

File: website/static/js/pages/render-nodes.js (hunk 2 above)
Comment (GageGaskins) [inline website/static/js/pages/render-nodes.js]: Fixed. Left this in on accident when testing to see if it would call only once.

File: website/static/js/pages/render-nodes.js (hunk 1 above)
Comment (GageGaskins) [inline website/static/js/pages/render-nodes.js]: Required for applyBindings cssSelector, line 464. If I pass the class name, it needs an element ID or jQuery will fail, trying to find something with '# .selector'. If I pass in the DOM element, jQuery fails with "can't find '[object Object] .selector'" or something like that.

File: website/templates/util/render_nodes.mako (hunk 3 above)
Comment (GageGaskins) [inline website/templates/util/render_nodes.mako]: Changed ID to camelCase as renderNodesList, class name to render-nodes-list for clarity.

**Comment by GageGaskins** [comment]
@lyndsysimon First pass changes complete

**Comment by lyndsysimon** [comment]
Looks great to me; I'm passing it on to Steve.

@sloria:

As background, this solution is a bit of a hack, but after talking about it neither I nor Gage came up with a better way. `website/static/js/pages/render-nodes.js` is getting included twice, so the logic to bind the viewmodel gets called twice. Gage added logic to detect if the element was already bound, and if so, to skip it.

If you can easily think of a better way to do this, please let us know. Otherwise, I see this is a minor hack that has been isolated as much as possible from the rest of the code.

**Comment by brianjgeiger** [comment]
Could you pass kwarg in through mod meta, or use the pluralized_node_type kwarg already there, to differentiate the id's used for scoping in the render_nodes.mako?

**Comment by GageGaskins** [comment]
New fix thanks to @brianjgeiger.

**Comment by sloria** [comment]
I've cherry-picked this into master since this is a critical fix. Thanks @GageGaskins

**Comment by lbanner** [comment]
:+1: TY
